File: mobile/src/main/java/org/openhab/habdroid/ui/widget/ChartScalingPreference.java
package org.openhab.habdroid.ui.widget;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.preference.Preference;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.SeekBar;
import android.widget.TextView;

import org.openhab.habdroid.R;

public class ChartScalingPreference extends Preference implements SeekBar.OnSeekBarChangeListener {
    private String[] mEntries;
    private float[] mValues;
    private SeekBar mSeekBar;
    private TextView mLabel;
    private float mValue;

    public ChartScalingPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public ChartScalingPreference(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    @TargetApi(21)
    public ChartScalingPreference(Context context, AttributeSet attrs,
            int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init();
    }

    @Override
    protected View onCreateView(ViewGroup parent) {
        View v = super.onCreateView(parent);
        mSeekBar = v.findViewById(R.id.seekbar);
        mSeekBar.setOnSeekBarChangeListener(this);
        mSeekBar.setMax(mValues.length - 1);
        for (int i = 0; i < mValues.length; i++) {
            if (mValues[i] == mValue) {
                mSeekBar.setProgress(i);
                break;
            }
        }

        mLabel = v.findViewById(R.id.label);
        updateLabel();

        return v;
    }

    private void init() {
        setLayoutResource(R.layout.chart_scaling_pref);

        Resources res = getContext().getResources();
        mEntries = res.getStringArray(R.array.chartScalingEntries);
        int[] values = res.getIntArray(R.array.chartScalingValues);
        mValues = new float[values.length];
        for (int i = 0; i < values.length; i++) {
            mValues[i] = (float) values[i] / 100f;
        }
    }

    @Override
    protected Object onGetDefaultValue(TypedArray a, int index) {
        return a.getFloat(index, 1.0f);
    }

    @Override
    protected void onSetInitialValue(boolean restorePersistedValue, Object defaultValue) {
        super.onSetInitialValue(restorePersistedValue, defaultValue);
        float defaultFloat = defaultValue instanceof Float ? (Float) defaultValue : 1.0f;
        mValue = restorePersistedValue ? getPersistedFloat(defaultFloat) : defaultFloat;
    }

    @Override
    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromTouch) {
        if (!fromTouch) {
            return;
        }
        float value = mValues[progress];
        if (callChangeListener(value)) {
            mValue = value;
            updateLabel();
            if (isPersistent()) {
                persistFloat(value);
            }
        }
    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {

    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {

    }

    private void updateLabel() {
        mLabel.setText(mEntries[mSeekBar.getProgress()]);
    }
}

File: mobile/src/main/java/org/openhab/habdroid/ui/widget/ChartScalingPreference.kt
package org.openhab.habdroid.ui.widget

import android.content.Context
import android.content.res.TypedArray
import android.preference.Preference
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.widget.SeekBar
import android.widget.TextView

import org.openhab.habdroid.R

class ChartScalingPreference constructor(context: Context, attrs: AttributeSet):
        Preference(context, attrs), SeekBar.OnSeekBarChangeListener {
    private val entries: Array<String>
    private val values: Array<Float>
    private lateinit var seekBar: SeekBar
    private lateinit var label: TextView
    private var value: Float = 0F

    init {
        layoutResource = R.layout.chart_scaling_pref

        val res = context.resources
        entries = res.getStringArray(R.array.chartScalingEntries)
        val intValues = res.getIntArray(R.array.chartScalingValues)
        values = intValues.map { v -> v.toFloat() / 100F }.toTypedArray()
    }

    override fun onCreateView(parent: ViewGroup): View {
        val view = super.onCreateView(parent)
        seekBar = view.findViewById(R.id.seekbar)
        seekBar.setOnSeekBarChangeListener(this)
        seekBar.max = values.size - 1
        seekBar.progress = Math.max(0, values.indexOfFirst { v -> v == value })

        label = view.findViewById(R.id.label)
        updateLabel()

        return view
    }

    override fun onGetDefaultValue(a: TypedArray, index: Int): Any {
        return a.getFloat(index, 1.0f)
    }

    override fun onSetInitialValue(restorePersistedValue: Boolean, defaultValue: Any?) {
        super.onSetInitialValue(restorePersistedValue, defaultValue)
        val defaultFloat = defaultValue as? Float ?: 1.0f
        value = if (restorePersistedValue) getPersistedFloat(defaultFloat) else defaultFloat
    }

    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromTouch: Boolean) {
        if (!fromTouch) {
            return
        }
        val value = values[progress]
        if (callChangeListener(value)) {
            this.value = value
            updateLabel()
            if (isPersistent) {
                persistFloat(value)
            }
        }
    }

    override fun onStartTrackingTouch(seekBar: SeekBar) {

    }

    override fun onStopTrackingTouch(seekBar: SeekBar) {

    }

    private fun updateLabel() {
        label.text = entries[seekBar.progress]
    }
}
